force-app/main/default/classes/AccountTriggerTest.cls

```diff
@@ -0,0 +1,245 @@
+        'Failed to get expected discount on child Accounts'
+      );
+    }
+  }
+
+  @isTest
+  private static void reparentingAccounts() {
+    List<Account> firstAccount = [
+      SELECT Id, Discount__c
+      FROM Account
+      LIMIT 1
+    ];
+    Id firstAccId = firstAccount[0].Id;
+
+    List<Account> childAccounts = createChildAccounts(firstAccount, false);
+    Account secondAccount = [
+      SELECT Id, Discount__c
+      FROM Account
+      WHERE Name LIKE 'Test Parent Account%' AND Id != :firstAccId
+      LIMIT 1
+    ];
+    Id secondAccId = secondAccount.Id;
+    secondAccount.Discount__c = DISCOUNT;
+    update secondAccount;
+
+    childAccounts = [
+      SELECT Id, Discount__c
+      FROM Account
+      WHERE Account.ParentId = :firstAccount
+    ];
+    for (Account c : childAccounts) {
+      c.ParentId = secondAccId;
+    }
+
+    Test.startTest();
+    update childAccounts;
+    Test.stopTest();
+    //count of child accounts related to first account
+    Assert.areEqual(
+      0,
+      [
+        SELECT COUNT()
+        FROM Account
+        WHERE ParentId = :firstAccId AND Discount__c = 20
+      ],
+      'Failed to get expected amount of child accounts and discount on first account'
+    );
+    /*count of child accounts related to second account
+     *(discount on current parent account
+     *should be =:DISCOUNT after update)
+     */
+    Assert.areEqual(
+      NUM_OF_ACCOUNTS,
+      [
+        SELECT COUNT()
+        FROM Account
+        WHERE ParentId = :secondAccId AND Discount__c = :DISCOUNT
+      ],
+      'Failed to get expected amount of child accounts and discount on current account'
+    );
+  }
+
+  @isTest
+  private static void deleteAndUndeleteAccounts() {
+    List<Account> parents = [
+      SELECT Id, Discount__c
+      FROM Account
+      LIMIT :NUM_OF_ACCOUNTS
+    ];
+
+    List<Account> childAccounts = createChildAccounts(parents, false);
+    Test.startTest();
+    delete parents;
+
+    childAccounts = [
+      SELECT Id, Discount__c
+      FROM Account
+      WHERE Id IN :childAccounts
+    ];
+    /*When deleting a parent account,
+     *child accounts' discount should be = 0
+     *(if the discount field is unlocked)
+     */
+    for (Account c : childAccounts) {
+      Assert.areEqual(
+        0,
+        c.Discount__c,
+        'Failed to get expected discount on child Accounts after delete'
+      );
+    }
+
+    undelete parents;
+    Test.stopTest();
+
+    childAccounts = [
+      SELECT Id, Discount__c
+      FROM Account
+      WHERE Id IN :childAccounts
+    ];
+    /*When undeleting a parent account,
+     *child accounts' discount should be = discount of the parent
+     *(if the discount field is unlocked)
+     */
+    for (Account c : childAccounts) {
+      Assert.areEqual(
+        20,
+        c.Discount__c,
+        'Failed to get expected discount on child Accounts after undelete'
+      );
+    }
+  }
+
+  private static List<Account> createChildAccounts(
```

Comment:
it makes sense to add ApexDocs and maybe also asserts with messages expressing whats being accomplished here

i also think this creates only 1 level of children, not several?

Reply:
Yes, that's right... unfortunately the logic of my brain does not coincide with that of Apex.